test/helpers/position_assertions.cc

```diff
@@ -1152,24 +1154,129 @@ shared_ptr<SymbolSearchAssertion> SymbolSearchAssertion::make(string_view filena
+ *    unmatchedSymbols => symbols with no matching assertion
+ */
+void matchAssertionsToSymbols(
+    string_view query, const vector<shared_ptr<SymbolSearchAssertion>> &assertions,
+    vector<shared_ptr<SymbolInformation>> &symbols,
+    const UnorderedMap<string, shared_ptr<core::File>> &sourceFileContents, string_view uriPrefix,
+    string_view errorPrefix,
+    vector<pair<shared_ptr<SymbolSearchAssertion>, shared_ptr<SymbolInformation>>> &assertionSymbolPairs,
+    vector<shared_ptr<SymbolSearchAssertion>> &unmatchedAssertions,
+    vector<shared_ptr<SymbolInformation>> &unmatchedSymbols) {
+    std::map<std::pair<std::string, int>, vector<shared_ptr<SymbolSearchAssertion>>> assertionsByLine;
     for (auto assertion : assertions) {
         auto location = assertion->getLocation(uriPrefix);
-        auto line = fmt::format("{}:{:{}}", location->uri, location->range->start->line, lineWidth);
-        assertionsByLine[line].push_back(assertion);
+        assertionsByLine[make_pair(location->uri, location->range->start->line)].push_back(assertion);
+    }
+    std::map<std::pair<std::string, int>, vector<shared_ptr<SymbolInformation>>> symbolsByLine;
+    for (auto &symbol : symbols) {
+        auto key = make_pair(symbol->location->uri, symbol->location->range->start->line);
+        symbolsByLine[key].push_back(symbol);
+        assertionsByLine[key]; // create entry in assertionsByLine, if none exists yet.
+    }
+    // Look line-by-line for unmatched symbols and assertions
+    for (auto entry : assertionsByLine) {
+        auto assertionsOnLine = entry.second;
+        auto &symbolsOnLine = symbolsByLine[entry.first];
+        auto [uri, line] = entry;
+        // Find assertions with no matching symbol
+        for (auto &assertion : assertionsOnLine) {
+            optional<shared_ptr<SymbolInformation>> matchingSymbol;
+            for (auto &symbol : symbols) {
+                if (!assertion->matches(uriPrefix, symbol)) {
+                    continue;
+                }
+                if (matchingSymbol.has_value()) {
+                    addFailureAtLocationWithSource(
+                        assertion->getLocation(uriPrefix), sourceFileContents, uriPrefix,
+                        fmt::format("More than one symbol matches assertion:\n {}\n"
+                                    "Found matching symbols:\n  {}\n  {}\n"
+                                    "Please refine your test to avoid ambiguity.",
+                                    assertion->toString(), matchingSymbol.value()->toJSON(true), symbol->toJSON(true)));
+                } else {
+                    matchingSymbol = symbol;
+                    assertionSymbolPairs.emplace_back(assertion, symbol);
+                }
+            }
+            if (!matchingSymbol.has_value()) {
+                unmatchedAssertions.push_back(assertion);
+            }
+        }
+        // Find symbols with no matching assertion
```

Comment:
(I still don't completely follow why we can't mark matched assertions in the above loop.)

Reply:
Now marked